File: core/src/com/unibo/s3/main_system/characters/steer/collisions/Box2dProxyDetectorsFactory.java
package com.unibo.s3.main_system.characters.steer.collisions;

import akka.actor.ActorRef;
import com.badlogic.gdx.ai.steer.Proximity;
import com.badlogic.gdx.ai.utils.RaycastCollisionDetector;
import com.badlogic.gdx.math.Vector2;
import com.unibo.s3.main_system.world.actors.Box2dRayCastCollisionDetectorProxy;
import com.unibo.s3.main_system.world.actors.Box2dSquareAABBProximityProxy;

/**
 * A factory of 'proxy' detectors, based on WorldActor instead of a World instance directly.
 *
 * @author mvenditto
 */
public class Box2dProxyDetectorsFactory implements DetectorsFactory<Vector2> {

    private ActorRef worldActor;

    public Box2dProxyDetectorsFactory(ActorRef worldActor) {
        this.worldActor = worldActor;
    }

    @Override
    public RaycastCollisionDetector<Vector2> newRaycastCollisionDetector() {
        return new Box2dRayCastCollisionDetectorProxy(worldActor);
    }

    @Override
    public Proximity<Vector2> newProximityDetector(float detectionRadius) {
        return new Box2dSquareAABBProximityProxy(null, worldActor, detectionRadius);
    }
}

File: core/src/com/unibo/s3/main_system/characters/steer/collisions/Box2dProxyDetectorsFactory.scala
package com.unibo.s3.main_system.characters.steer.collisions

import akka.actor.{ActorRef, ActorSelection}
import com.badlogic.gdx.ai.steer.Proximity
import com.badlogic.gdx.ai.utils.RaycastCollisionDetector
import com.badlogic.gdx.math.Vector2
import com.unibo.s3.main_system.world.actors.ActorRefOrSelection._
import com.unibo.s3.main_system.world.actors.{Box2dRayCastCollisionDetectorProxy, Box2dSquareAABBProximityProxy}

/**
  * A factory class of Box2d detectors working with (remote too) actors.
  * @param ars an [[ActorRefOrSelectionHolder]] of the 'world actor'
  */
class Box2dProxyDetectorsFactory(ars: ActorRefOrSelectionHolder) extends DetectorsFactory[Vector2] {

  def newRaycastCollisionDetector(): RaycastCollisionDetector[Vector2] =
    new Box2dRayCastCollisionDetectorProxy(ars)

  def newProximityDetector(detectionRadius: Float): Proximity[Vector2] =
    new Box2dSquareAABBProximityProxy(null, ars, detectionRadius)
}

object Box2dProxyDetectorsFactory {
  def apply (ars: ActorRefOrSelectionHolder): Box2dProxyDetectorsFactory =
    new Box2dProxyDetectorsFactory(ars)

  def of(ar: ActorRef): Box2dProxyDetectorsFactory =
    Box2dProxyDetectorsFactory(ar)

  def of(as: ActorSelection): Box2dProxyDetectorsFactory =
    Box2dProxyDetectorsFactory(as)
}


